torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -212,6 +212,97 @@ void InlineFunctionalGraphs(Block* block) {
   }
 }
 
+struct MutationRemover {
+  MutationRemover(const std::shared_ptr<Graph>& graph)
+      : aliasDb_(nullptr), graph_(graph) {
+    aliasDb_ = torch::make_unique<AliasDb>(graph_);
+  }
+
+  void run() {
+    RemoveAtenMutation(graph_->block());
+  }
+
+ private:
+  bool uniqueAlias(Value* v) {
+    // if the output isn't contained or alias by the inputs to its node, it's
+    // unique
+    bool unique_alias = !aliasDb_->mayContainAlias(v->node()->inputs(), v);
+    // bail on nodes with side effects like prim::If etc
+    return unique_alias && !v->node()->hasSideEffects();
+  }
+
+  bool inplaceOpVariant(Node* n) {
+    if (!n->kind().is_aten()) {
+      return false;
+    }
+    auto name = n->schema().name();
+    bool inplace_op = name.at(name.size() - 1) == '_';
+    if (!inplace_op) {
+      return false;
+    }
+    auto new_schema = name.substr(0, name.size() - 1);
+    return getAllOperatorsFor(Symbol::fromQualString(new_schema)).size() != 0;
+  }
+
+  void RemoveAtenMutation(Block* block) {
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      auto* node = *it;
+      it++;
+
+      for (Block* sub_block : node->blocks()) {
+        RemoveAtenMutation(sub_block);
+      }
+
+      // TODO: out op variants
+      if (!inplaceOpVariant(node)) {
+        continue;
+      }
+
+      Value* mutated_value = node->inputs().at(0);
+
+      // We can only remove mutation to values that are unique aliases in the
```

Comment:
I think the concept of `unique` introduces confusion—I would expect it to mean "this value is the only pointer to this tensor memory" (a la `unique_ptr`). But in this case, if I had:
```
z = add(x, y)
a = z[0]
add_(z, …)
```
It appears that `uniqueAlias(z)` would return true, even though `a` points to `z`. 

It seems the rule we want is: 
1. We witnessed the creation of `foo`'s memory and know that it isn't an alias of anything. In implementation terms, `foo` is a `MemoryLocation` (a leaf on the points-to graph).
2. We can "fold" the in-place add into the creation of `foo`. In implementation terms, we can move `add_` after `foo->node()` in a topologically valid way.

Can we modify this pass to check `MemoryLocation` instead of using `uniqueAlias`?

Also: it would be helpful to have a high-level comment describing the rules. The comments here are good but somewhat in the weeds (they describe what we *can't* do in the name of safety, but don't present the overall model of what we *want* to do, if that makes any sense…)

Reply:
I don't think the case you've added here would actually induces a changes in behavior. This would fail because the creation and subsequent mutation could not be made atomic. 

Sure, i can add more description & high-level commentary.

Reply:
I'll look into using the MemoryLocation here

Reply:
What you've described in 1 and 2 is the same as what already exists, using `MemoryLocations` instead of the aliasDb to check if an op is the creation of a tensor..

Reply:
Actually, I take that back. You've added the additional check that: `"this value is the only pointer to this tensor memory" (a la unique_ptr)`, however this is an unnecessary restriction.

Reply:
Sorry I should be a little clearer: I think what I've suggested is the same functionally as what you have here, it just re-uses an existing concept (`MemoryLocation`) instead of adding a new one ("uniqueness"). So really it's just shuffling around concepts to make the code a little clearer (to me at least).

When I was talking about "this value is the only tensor memory"—I'm not saying that should be the requirement, I'm just saying that that's what I assume "unique" means given what it means in other contexts, which is confusing because that's not precisely what we want here, as you've pointed out.

Reply:
So concretely, the change I am suggesting is: rewrite using an `isMemoryLocation` function instead of `uniqueAlias`.